crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,185 @@
+        save_local_interfaces(&get_if_addr);
+
+        let if_cache_updater_handle = Some(tokio::spawn({
+            let mut notify = PATH_CHANGE_BROADCAST.subscribe();
+            async move {
+                loop {
+                    match notify.recv().await {
+                        Ok(()) => save_local_interfaces(&get_if_addr),
+                        Err(e) => {
+                            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
+                        }
+                    }
+                }
+            }
+        }));
+
+        // Only add in production, otherwise network triggers from OS can occur
+        // which can make the test (test_path_change_notification) flaky sometimes
+        #[cfg(all(
+            not(test),
+            any(target_os = "macos", target_os = "ios", target_os = "tvos")
+        ))]
+        NW_PATH_MONITOR_START.call_once(crate::apple::setup_network_monitor);
+        #[cfg(target_os = "linux")]
+        let monitor_handle = crate::linux::setup_network_monitor().await;
+
+        #[cfg(target_os = "windows")]
+        let monitor_handle = crate::windows::setup_network_monitor();
+
+        Ok(Self {
+            if_cache_updater_handle,
+            #[cfg(any(target_os = "linux", target_os = "windows"))]
+            monitor_handle,
+        })
+    }
+}
+
+impl Drop for NetworkMonitor {
+    fn drop(&mut self) {
+        if let Some(handle) = &self.if_cache_updater_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "linux")]
+        if let Some(handle) = &self.monitor_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "windows")]
+        if !self.monitor_handle.0.is_null() {
+            crate::windows::deregister_network_monitor(self.monitor_handle.clone());
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::MockGetIfAddrs;
+
+    use super::*;
+    use serial_test::serial;
+    use std::{
+        net::Ipv4Addr,
+        sync::atomic::{AtomicU8, Ordering},
+    };
+
+    static CALL_COUNT: AtomicU8 = AtomicU8::new(0);
+
+    async fn setup_network_monitor() -> NetworkMonitor {
+        CALL_COUNT.store(0, Ordering::SeqCst);
+        let mut get_if_addrs_mock = MockGetIfAddrs::new();
+        get_if_addrs_mock.expect_get().returning(|| {
+            CALL_COUNT.fetch_add(1, Ordering::SeqCst);
+            Ok(vec![if_addrs::Interface {
+                name: "correct".to_owned(),
+                addr: if_addrs::IfAddr::V4(if_addrs::Ifv4Addr {
+                    ip: Ipv4Addr::new(10, 0, 0, 1),
+                    netmask: Ipv4Addr::new(255, 255, 255, 0),
+                    broadcast: None,
+                }),
+                index: None,
+                #[cfg(windows)]
+                adapter_name: "{78f73923-a518-4936-ba87-2a30427b1f63}".to_string(),
+            }])
+        });
+        NetworkMonitor::new(get_if_addrs_mock).await.unwrap()
+    }
+
+    #[tokio::test]
+    #[serial]
```

Comment:
you can discard the serial crate and make them run serially with a mutex:
```suggestion
static TESTS_SYNC_MTX: Mutex<()> = Mutex::new(());

    #[tokio::test]
    async fn test_path_change_notification() {
        let _mut = TESTS_SYNC_MTX.lock();
```

Reply:
(optionally)